expression/builtin_miscellaneous.go

```diff
@@ -79,55 +78,52 @@ type sleepFunctionClass struct {
 
 func (c *sleepFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
 	err := errors.Trace(c.verifyArgs(args))
-	bt := &builtinSleepSig{newBaseBuiltinFunc(args, ctx)}
-	bt.deterministic = false
-	return bt.setSelf(bt), errors.Trace(err)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpInt)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinSleepSig{baseIntBuiltinFunc{bf}}
+	sig.deterministic = false
+	return sig.setSelf(sig), errors.Trace(err)
 }
 
 type builtinSleepSig struct {
-	baseBuiltinFunc
+	baseIntBuiltinFunc
 }
 
-// eval evals a builtinSleepSig.
+// evalInt evals a builtinSleepSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html#function_sleep
-func (b *builtinSleepSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+func (b *builtinSleepSig) evalInt(row []types.Datum) (int64, bool, error) {
+	val, isNull, err := b.args[0].EvalInt(row, b.ctx.GetSessionVars().StmtCtx)
 	if err != nil {
-		return types.Datum{}, errors.Trace(err)
+		return 0, isNull, errors.Trace(err)
 	}
 	sessVars := b.ctx.GetSessionVars()
-	if args[0].IsNull() {
+	if isNull {
 		if sessVars.StrictSQLMode {
```

Comment:
I think separation is trivial since it won't bring us significant performance improvement for this function (only differs with 1 logic test statement) but make the code much more non-intuitive & harder to read.